tests/utest_memory_usage.cpp

```diff
@@ -0,0 +1,267 @@
+  size_t expected_size = 2409;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add a layer.
+  graph.addLayer(1, 0, "test_layer");
+  expected_size += 458;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add a partitioned layer.
+  graph.addLayer(2, 1, "test_layer_partition");
+  expected_size += 120;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add some nodes.
+  for (size_t i = 0; i < 20; ++i) {
+    graph.emplaceNode(1, i, std::make_unique<NodeAttributes>(), 0);
+    graph.emplaceNode(2, 20 + i, std::make_unique<NodeAttributes>(), 1);
+  }
+  expected_size += 6140;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add some intralayer edges.
+  for (size_t i = 0; i < 20; ++i) {
+    graph.insertEdge(i, (i + 1) % 20);
+    graph.insertEdge(20 + i, 20 + (i + 1) % 20);
+  }
+  expected_size += 40 * 58;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add some interlayer edges.
+  for (size_t i = 0; i < 20; ++i) {
+    graph.insertEdge(i, 20 + i);
+  }
+  expected_size += 20 * 108;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add mesh.
+  auto mesh = std::make_unique<Mesh>();
+  mesh->points.resize(1000);
+  mesh->colors.resize(1000);
+  graph.setMesh(std::move(mesh));
+  expected_size += 28160;  // Mesh size.
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+
+  // Add metadata.
+  nlohmann::json meta_contents;
+  meta_contents["creator"] = "unit_test";
+  meta_contents["description"] = "Test scene graph";
+  graph.metadata.set(meta_contents);
+  expected_size += 56;
+  EXPECT_EQ(graph.memoryUsage(), expected_size);
+}
+
+SceneGraph::Ptr populateSceneGraph(size_t size) {
+  auto graph = std::make_shared<SceneGraph>();
+  graph->addLayer(1, 0, "test_layer");
+  graph->addLayer(2, 1, "test_layer_partition");
+  auto attrs = std::make_unique<SemanticNodeAttributes>();
+  attrs->semantic_feature.resize(128, 64);
+  for (size_t i = 0; i < size; ++i) {
+    graph->emplaceNode(1, i, attrs->clone(), 0);
+    graph->emplaceNode(2, size + i, attrs->clone(), 1);
+  }
+  for (size_t i = 0; i < size; ++i) {
+    graph->insertEdge(i, (i + 1) % size);
+    graph->insertEdge(size + i, size + (i + 1) % size);
+  }
+  for (size_t i = 0; i < size; ++i) {
+    graph->insertEdge(i, size + i);
+  }
+  auto mesh = std::make_unique<Mesh>();
+  mesh->points.resize(10 * size);
+  mesh->colors.resize(10 * size);
+  graph->setMesh(std::move(mesh));
+  return graph;
+}
+
+TEST(MemoryUsage, SerializationComparison) {
+  for (const auto& size : {10, 100, 1000}) {
+    // Serialize to a byte stream and compare with estimated memory usage.
+    SCOPED_TRACE("Scene graph size: " + std::to_string(size));
+    auto graph = populateSceneGraph(size);
+    std::vector<uint8_t> buffer;
+    io::binary::writeGraph(*graph, buffer, true);
+    const double ratio =
+        static_cast<double>(buffer.size()) / static_cast<double>(graph->memoryUsage());
+    EXPECT_LT(ratio, 2.0);
+    EXPECT_GT(ratio, 1.9);
```

Comment:
(minor) I'd maybe make this just `EXPECT_GT(ratio, 1.0)` to avoid having to update the unit test if we update how we compute memory (e.g., if we exactly compute the node attributes memory requirements)